src/dataflow/src/render/mod.rs

```diff
@@ -749,6 +749,7 @@ where
                 collection
             }
             SinkEnvelope::Tail { emit_progress } => keyed
+                .consolidate()
```

Comment:
I'm a little bit confused because I thought the `arrange_by_key` in line 817 would consolidate updates and am not entirely sure why that doesn't work

Reply:
Is it because the key is the sink ID? It seems reasonable that `arrange_by_key` would only cause updates to be consolidated by the key.

Reply:
ohhh wait i think i get it. line 761 adds the diff to the row and line 764 turns everything into a positive diff

so previously, we might see two records like

```
(foo, t1, +1)
(foo, t1, -1)
```

and its clear in that instance that we are modifying the same data at the same time so we can consolidate the update. but the map turns this into

```
([foo, t1, +1], t1, 1)
([foo, t1, -1], t1, 1)
```
If this theory is correct then adding the consolidate at any point after the map would fail to resolve the issue

Reply:
Is there a reason that we add the `(_, t1, 1)` and cast things to a `Collection`? It seems like that information is now at best duplicated / redundant, and worst could be misinterpreted elsewhere. Should it just be a `Stream<G, Row>`?

Reply:
I agree that seems weird.

Reply:
The logic looks a bit spaghetti at the moment. It seems like there are a sequence of match statements for each of the flavors of sink, which e.g. push the timestamp for Tail, then bounce out and consult the `append_timestamp` bool to see if they should do it again, then match on connector again. It's probably not for a reason and just some accretion of complexity.

Reply:
Maybe just a design decision too; one could append the time and diff in the `sink::tail` implementation, where we currently do
```
    for _ in 0..diff {
        // Add the unpacked timestamp so we can sort by them later.
        results.push((*time, row.clone()));
    }
```
but instead push the packing action back out to the prep work. It's def an anti-pattern (if anyone were holding on to the arrangement, which they are not, it would result in an arrangement that cannot compact itself, because we baked times and diffs into the data and they will not cancel (as seen above).

Reply:
I'm working on it. PR incoming!

Reply:
I think maybe what happened is some crossed wires. A while back, there was an effort to make sinks generally consume pre-arranged batches of data, so that sinking or tailing materializations wouldn't require a new copy of the data. That bit seems to have stuck, but rather than push e.g. formatting in to the sink implementation, it was pushed upstream which undoes the value of defining the operators in terms of arranged batches.

Reply:
@frankmcsherry moving the timestamp and diff into the data made sense to me as part of moving "envelope" logic out of the tail connector implementations. I.e. tail.rs now just dumbly prints out whatever data you give it, and doesn't need to know that ts/diff are expected columns in the output now.